subsys/bluetooth/audio/shell/cap_initiator.c

```diff
@@ -481,14 +484,675 @@ static int cmd_cap_initiator_unicast_cancel(const struct shell *sh, size_t argc,
+	size_t snk_cnt = 0;
+	size_t src_cnt = 0;
+	int err;
+
+	if (default_unicast_group != NULL) {
+		shell_error(sh, "Unicast Group already exist, please delete first");
+		return -ENOEXEC;
+	}
+
+	if (param->conn_cnt > BAP_UNICAST_AC_MAX_CONN) {
+		shell_error(sh, "Invalid conn_cnt: %zu", param->conn_cnt);
+		return -ENOEXEC;
+	}
+
+	for (size_t i = 0; i < param->conn_cnt; i++) {
+		/* Verify conn values */
+		if (param->snk_cnt[i] > BAP_UNICAST_AC_MAX_SNK) {
+			shell_error(sh, "Invalid conn_snk_cnt[%zu]: %zu", i, param->snk_cnt[i]);
+			return -ENOEXEC;
+		}
+
+		if (param->src_cnt[i] > BAP_UNICAST_AC_MAX_SRC) {
+			shell_error(sh, "Invalid conn_src_cnt[%zu]: %zu", i, param->src_cnt[i]);
+			return -ENOEXEC;
+		}
+	}
+
+	/* Populate the array of connected connections */
+	bt_conn_foreach(BT_CONN_TYPE_LE, populate_connected_conns, (void *)connected_conns);
+	for (conn_avail_cnt = 0; conn_avail_cnt < ARRAY_SIZE(connected_conns); conn_avail_cnt++) {
+		if (connected_conns[conn_avail_cnt] == NULL) {
+			break;
+		}
+	}
+
+	if (conn_avail_cnt < param->conn_cnt) {
+		shell_error(sh,
+			    "Only %zu/%u connected devices, please connect additional devices for "
+			    "this audio configuration",
+			    conn_avail_cnt, param->conn_cnt);
+		return -ENOEXEC;
+	}
+
+	/* Set all endpoints from multiple connections in a single array, and verify that the known
+	 * endpoints matches the audio configuration
+	 */
+	for (size_t i = 0U; i < param->conn_cnt; i++) {
+		for (size_t j = 0U; j < param->snk_cnt[i]; j++) {
+			snk_cnt++;
+		}
+
+		for (size_t j = 0U; j < param->src_cnt[i]; j++) {
+			src_cnt++;
+		}
+	}
+
+	if (snk_cnt > 0U) {
+		snk_named_preset = bap_get_named_preset(true, argv[1]);
+		if (snk_named_preset == NULL) {
+			shell_error(sh, "Unable to parse snk_named_preset %s", argv[1]);
+			return -ENOEXEC;
+		}
+	}
+
+	if (src_cnt > 0U) {
+		const char *preset_arg = argc > 2 ? argv[2] : argv[1];
+
+		src_named_preset = bap_get_named_preset(true, preset_arg);
+		if (src_named_preset == NULL) {
+			shell_error(sh, "Unable to parse src_named_preset %s", argv[1]);
+			return -ENOEXEC;
+		}
+	}
+
+	if (!ctx_shell) {
+		ctx_shell = sh;
+	}
+
+	/* Setup arrays of parameters based on the preset for easier access. This also copies the
+	 * preset so that we can modify them (e.g. update the metadata)
+	 */
+	for (size_t i = 0U; i < snk_cnt; i++) {
+		snk_uni_streams[i] = &unicast_streams[i];
+
+		if (unicast_streams[i].stream.bap_stream.conn != NULL) {
+			shell_error(sh, "unicast_streams[%zu] already in use", i);
+			return -ENOEXEC;
+		}
+
+		copy_unicast_stream_preset(snk_uni_streams[i], snk_named_preset);
+
+		/* Some audio configuration requires multiple sink channels,
+		 * so multiply the SDU based on the channel count
+		 */
+		snk_uni_streams[i]->qos.sdu *= param->snk_chan_cnt;
+	}
+
+	for (size_t i = 0U; i < src_cnt; i++) {
+		src_uni_streams[i] = &unicast_streams[i + snk_cnt];
+
+		if (unicast_streams[i].stream.bap_stream.conn != NULL) {
```

Comment:
should this have been src_uni_streams[i] ? 

also, just curious - are we sure the sinks will always before the sources?

Reply:
> also, just curious - are we sure the sinks will always before the sources?

We decide this :) I have considered splitting the `unicast_streams` into an array for sinks and an array for sources

Reply:
> > also, just curious - are we sure the sinks will always before the sources?
> 
> We decide this :) I have considered splitting the `unicast_streams` into an array for sinks and an array for sources

Splitting would make it more clear, imo.  But not critical (if not done now, then maybe add a comment where the split is assumed inside the same array)